### PR TITLE
Added device model type definition for Bluetooth protocol in device CRD design

### LINKED ISSUE #361: Add support for bluetooth protocol in Device CRD design
**What would you like to be added**:
Device CRD design should contain device model type definition for Bluetooth protocol.

**Why is this needed**:
Device model type definition for Bluetooth protocol is necessary to create a mapper which uses device CRD to provide support for Bluetooth devices

### PR DESCRIPTION
**Pull request by sujithsimon22**
**What type of PR is this?**
 /kind design

**What this PR does / why we need it**:
This PR provides device model type definition for Bluetooth protocol in the device CRD design

**Which issue(s) this PR fixes**:

Fixes #361 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by sujithsimon22** [comment]
cc @sids-b @rohitsardesai83 @m1093782566  Pls review, thnks :)

**Comment by m1093782566** [comment]
cc @qizha @kevin-wangzefeng @zhenghaibo1008

File: docs/proposals/device-crd.md [hunk 1]
```diff
@@ -185,8 +185,55 @@ type VisitorConfig struct {
...
+	//+optional
+	OrderOfOperations []map[ArithmaticOperationType]float64 `json:"order-of-operations"`
+}
+
+// Operations supported by Bluetooth protocol to convert the value being read from the device into an understandable form
+type ArithmaticOperationType string
+
+// Bluetooth Protocol Operation type
+const (
+	Add      ArithmaticOperationType = "Add"
```
Comment (m1093782566) [inline docs/proposals/device-crd.md]: If these operations are only for Bluetooth, please add the `Bluetooth` prefix.

Comment (sujithsimon22) [reply docs/proposals/device-crd.md]: Makes sense, added as requested. Thanks :)

File: docs/proposals/device-crd.md [hunk 2]
```diff
@@ -185,8 +185,55 @@ type VisitorConfig struct {
...
+}
+
+// Common visitor configurations for bluetooth protocol
+type VisitorConfigBluetooth struct {
+	// Required: Unique ID of the corresponding operation
+	CharacteristicUUID string `json:"characteristic-uuid,omitempty"`
+	//Responsible for converting the data coming from the platform into a form that is understood by the bluetooth device
+	// For example: "ON":[1], "OFF":[0]
+	//+optional
+	BluetoothDataWrite map[string][]byte `json:"write-data,omitempty"`
```
Comment (m1093782566) [inline docs/proposals/device-crd.md]: Camel naming please.

File: docs/proposals/device-crd.md [hunk 3]
```diff
@@ -185,8 +185,55 @@ type VisitorConfig struct {
...
+	// Responsible for converting the data being read from the bluetooth device into a form that is understandable by the platform
+	//+optional
+	BluetoothDataRead BluetoothDataRead `json:"read-data,omitempty"`
```
Comment (m1093782566) [inline docs/proposals/device-crd.md]: Camel naming please.

File: docs/proposals/device-crd.md [hunk 4]
```diff
@@ -185,8 +185,55 @@ type VisitorConfig struct {
...
 }
 
+// Specifies the operations that may need to be performed to convert the data
+type BluetoothDataRead struct {
+	// Required: Specifies the start index of the incoming byte stream to be considered to convert the data.
+	// For example: start-index:2, end-index:3 concatenates the value present at second and third index of the incoming byte stream. If we want to reverse the order we can give it as start-index:3, end-index:2
+	StartIndex int `json:"start-index"`
```
Comment (m1093782566) [inline docs/proposals/device-crd.md]: Camel naming here and below.

File: docs/proposals/device-crd.md (hunk 1 above)
Comment (m1093782566) [reply docs/proposals/device-crd.md]: Please add `Bluetooth` prefix for `ArithmaticOperationType` as well.

File: docs/proposals/device-crd.md (hunk 2 above)
Comment (sujithsimon22) [reply docs/proposals/device-crd.md]: @m1093782566 sry, i didn't get it....."BluetoothDataWrite" is this exported variable not in camel case already?

Comment (m1093782566) [reply docs/proposals/device-crd.md]: I mean "write-data"

Comment (sujithsimon22) [reply docs/proposals/device-crd.md]: oh...got it, thanks. Done

File: docs/proposals/device-crd.md (hunk 4 above)
Comment (sujithsimon22) [reply docs/proposals/device-crd.md]: done, thanks

File: docs/proposals/device-crd.md (hunk 3 above)
Comment (sujithsimon22) [reply docs/proposals/device-crd.md]: done, thanks

File: docs/proposals/device-crd.md [hunk 5]
```diff
@@ -185,8 +185,55 @@ type VisitorConfig struct {
...
+	// +optional
+	ShiftRight uint `json:"shiftRight"`
+	// Specifies the operations that are required to be performed, with what value the operation is to be performed and in what order they are to be performed
+	// For example: ["Multiply":0.235, "Add":32] will ensure that the value is multiplied by 0.235 and the result is added with 32 to provide the final result
+	//+optional
+	OrderOfOperations []map[BluetoothArithmaticOperationType]float64 `json:"orderOfOperations"`
+}
+
+// Operations supported by Bluetooth protocol to convert the value being read from the device into an understandable form
+type BluetoothArithmaticOperationType string
```
Comment (rohitsardesai83) [inline docs/proposals/device-crd.md]: Arithmatic -> Arithmetic

File: docs/proposals/device-crd.md [hunk 6]
```diff
@@ -185,8 +185,55 @@ type VisitorConfig struct {
...
+
+// Bluetooth Protocol Operation type
+const (
+	BluetoothAdd      BluetoothArithmaticOperationType = "Add"
```
Comment (rohitsardesai83) [inline docs/proposals/device-crd.md]: Arithmatic -> Arithmetic

File: docs/proposals/device-crd.md [hunk 7]
```diff
@@ -185,8 +185,55 @@ type VisitorConfig struct {
...
+	// Refers to the number of bits to shift left, if left-shift operation is necessary for conversion
+	// +optional
+	ShiftLeft uint `json:"shiftLeft"`
+	// Refers to the number of bits to shift right, if right-shift operation is necessary for conversion
+	// +optional
+	ShiftRight uint `json:"shiftRight"`
+	// Specifies the operations that are required to be performed, with what value the operation is to be performed and in what order they are to be performed
+	// For example: ["Multiply":0.235, "Add":32] will ensure that the value is multiplied by 0.235 and the result is added with 32 to provide the final result
+	//+optional
+	OrderOfOperations []map[BluetoothArithmaticOperationType]float64 `json:"orderOfOperations"`
```
Comment (rohitsardesai83) [inline docs/proposals/device-crd.md]: Arithmatic -> Arithmetic

File: docs/proposals/device-crd.md (hunk 5 above)
Comment (sujithsimon22) [reply docs/proposals/device-crd.md]: Done, thanks

File: docs/proposals/device-crd.md (hunk 7 above)
Comment (sujithsimon22) [reply docs/proposals/device-crd.md]: done, thanks

File: docs/proposals/device-crd.md (hunk 6 above)
Comment (sujithsimon22) [reply docs/proposals/device-crd.md]: done, thanks

**Comment by sujithsimon22** [comment]
> I think you will need to update the sample Device yaml file with bluetooth added.

I have added bluetooth specific information into the device model sample. Thanks

File: docs/proposals/device-crd.md [hunk 8]
```diff
@@ -185,8 +185,62 @@ type VisitorConfig struct {
...
 	Modbus VisitorConfigModbus `json:"modbus,omitempty"`
+	// Bluetooth represents a set of additional visitor config fields of bluetooth protocol.
+	// +optional
+	Bluetooth VisitorConfigBluetooth `json:"bluetooth,omitempty"`
+}
+
+// Common visitor configurations for bluetooth protocol
+type VisitorConfigBluetooth struct {
+	// Required: Unique ID of the corresponding operation
+	CharacteristicUUID string `json:"characteristicUuid,omitempty"`
```
Comment (kevin-wangzefeng) [inline docs/proposals/device-crd.md]: s/characteristicUuid/characteristicUUID/

File: docs/proposals/device-crd.md [hunk 9]
```diff
@@ -185,8 +185,62 @@ type VisitorConfig struct {
...
+	//Responsible for converting the data coming from the platform into a form that is understood by the bluetooth device
```
Comment (kevin-wangzefeng) [inline docs/proposals/device-crd.md]: `//Responsible` => `// Responsible`

File: docs/proposals/device-crd.md [hunk 10]
```diff
@@ -292,6 +358,21 @@ spec:
       scale: 1.0
       isSwap: true
       isRegisterSwap: true
+  - propertyName: pressure-enable
+    bluetooth:
+      characteristicUuid: f000aa4204514000b000000000000000
```
Comment (kevin-wangzefeng) [inline docs/proposals/device-crd.md]: s/characteristicUuid/characteristicUUID/

File: docs/proposals/device-crd.md [hunk 11]
```diff
@@ -292,6 +358,21 @@ spec:
...
+      dataWrite:
+        ON: [1]
+        OFF: [0]
+  - propertyName: pressure
+    bluetooth:
+      characteristicUuid: f000aa4104514000b000000000000000
```
Comment (kevin-wangzefeng) [inline docs/proposals/device-crd.md]: s/characteristicUuid/characteristicUUID/

File: docs/proposals/device-crd.md (hunk 8 above)
Comment (sujithsimon22) [reply docs/proposals/device-crd.md]: Done, thanks

File: docs/proposals/device-crd.md (hunk 9 above)
Comment (sujithsimon22) [reply docs/proposals/device-crd.md]: Done, thanks

File: docs/proposals/device-crd.md (hunk 10 above)
Comment (sujithsimon22) [reply docs/proposals/device-crd.md]: done, thanks

File: docs/proposals/device-crd.md (hunk 11 above)
Comment (sujithsimon22) [reply docs/proposals/device-crd.md]: done, thanks for the review :)

**Comment by kevin-wangzefeng** [comment]
We shall try speed up our CI by skipping e2e tests when the change obviously not relevant to that.

**Comment by m1093782566** [comment]
> We shall try speed up our CI by skipping e2e tests when the change obviously not relevant to that.

Okay, will add this task.

**Pull request merged**